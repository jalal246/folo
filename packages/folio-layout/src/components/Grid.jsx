import React from "react";
import PropTypes from "prop-types";

import {
  GRID,
  STRETCH,
  CENTER,
  SPACE_BETWEEN,
  AUTO,
  FR,
  AUTO_FIT,
  DEFAULT_GAP
} from "../constants";

const propTypes = {
  /**
   * custom render-component
   */
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

  /**
   * number of columns in grid
   */
  col: PropTypes.number,
  colWidth: PropTypes.string,
  colMinWidth: PropTypes.string,
  colMaxWidth: PropTypes.string,

  /**
   * number of rows in grid
   */
  row: PropTypes.number,
  rowWidth: PropTypes.string,
  rowMinWidth: PropTypes.string,
  rowMaxWidth: PropTypes.string,

  style: PropTypes.objectOf(PropTypes.string),
  //
  children: PropTypes.node.isRequired
};

const defaultProps = {
  component: "div",

  col: null,
  colWidth: undefined,
  colMinWidth: AUTO,
  colMaxWidth: FR,

  row: null,
  rowWidth: undefined,
  rowMinWidth: AUTO,
  rowMaxWidth: FR,
  style: {}
};

/**
 * call repeat() CSS function
 * depending on length, min,max
 * with some enhancements
 *
 *
 * @param {number} length row or column number in grid
 * @param {string} min minimum unit for row or column
 * @param {string} max maximum unit for row or column
 * @return {string}   repeat() CSS function represents a repeated fragment of the track list
 */
function repeat(length, fixed, min, max) {
  return `repeat(${length}, ${fixed || `minmax(${min}, ${max})`})`;
}

class Grid extends React.PureComponent {
  render() {
    const {
      component: CellComponent,

      col,
      colWidth,
      colMinWidth,
      colMaxWidth,

      row,
      rowWidth,
      rowMinWidth,
      rowMaxWidth,

      isCenter,

      // TODO: add style and shape to propTypes
      style: {
        display = GRID,

        alignItems = isCenter ? CENTER : STRETCH,

        // eslint-disable-next-line
        justifyContent = col || colWidth
          ? SPACE_BETWEEN
          : isCenter
            ? CENTER
            : STRETCH,

        gap = DEFAULT_GAP,

        ...otherStyles
      },
      children,

      //
      ...otherProps
    } = this.props;

    const style = {
      display,

      ...(row
        ? { gridTemplateRows: repeat(row, rowWidth, rowMinWidth, rowMaxWidth) }
        : rowWidth && { gridAutoRows: rowWidth }),

      ...(colWidth && !col
        ? { gridAutoColumns: colWidth }
        : {
            gridTemplateColumns: repeat(
              col || AUTO_FIT,
              colWidth,
              colMinWidth,
              colMaxWidth
            )
          }),

      alignItems,
      justifyContent,
      gap,
      ...otherStyles
    };

    return (
      <CellComponent style={style} {...otherProps}>
        {children}
      </CellComponent>
    );
  }
}

Grid.propTypes = propTypes;
Grid.defaultProps = defaultProps;

export default Grid;
