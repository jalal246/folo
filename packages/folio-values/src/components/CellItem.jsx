import React from "react";
import PropTypes from "prop-types";

// import componentShape from "../shapes/componentShape";

// import { keyGenerator } from "../../utils";

const propTypes = {
  /**
   * custom render-component
   */
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  /**
   * @see See [key](https://reactjs.org/docs/lists-and-keys.html) for a list of color names
   */
  // key: PropTypes.string,
  children: PropTypes.node.isRequired
};

const defaultProps = {
  component: "option"
};

/**
 * Used only when the Cell type is list
 * render option as defautl
 */
export default function CellItem({
  component: CellComponent,
  children,
  ...other
}) {
  return <CellComponent {...other}>{children}</CellComponent>;
}

CellItem.propTypes = propTypes;
CellItem.defaultProps = defaultProps;
