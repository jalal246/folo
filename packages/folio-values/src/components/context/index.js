export { ValuesConsumer, ValuesProvider } from "./ValuesContext";
