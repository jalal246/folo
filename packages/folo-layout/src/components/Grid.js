import React from "react";

import {
  GRID,
  STRETCH,
  CENTER,
  SPACE_BETWEEN,
  AUTO,
  FR,
  AUTO_FIT,
  DEFAULT_GAP,
} from "../constants";

/**
 * call repeat() CSS function
 * depending on length, min,max
 * with some enhancements
 *
 *
 * @param {number} length row or column number in grid
 * @param {string} min minimum unit for row or column
 * @param {string} max maximum unit for row or column
 * @return {string}   repeat() CSS function represents a repeated fragment of the track list
 */
function repeat(length, fixed, min, max) {
  return `repeat(${length}, ${fixed || `minmax(${min}, ${max})`})`;
}

const Grid = (props) => {
  const {
    component: CellComponent = "div",

    col,
    colWidth,
    colMinWidth = AUTO,
    colMaxWidth = FR,

    row,
    rowWidth,
    rowMinWidth,
    rowMaxWidth,

    isCenter = false,

    style: {
      display = GRID,

      alignItems = isCenter ? CENTER : STRETCH,

      // eslint-disable-next-line
      justifyContent = col || colWidth
        ? SPACE_BETWEEN
        : isCenter
        ? CENTER
        : STRETCH,

      gap = DEFAULT_GAP,

      ...otherStyles
    } = {},
    children,

    //
    ...rest
  } = props;

  const style = {
    display,

    ...(row
      ? { gridTemplateRows: repeat(row, rowWidth, rowMinWidth, rowMaxWidth) }
      : rowWidth && { gridAutoRows: rowWidth }),

    ...(colWidth && !col
      ? { gridAutoColumns: colWidth }
      : {
          gridTemplateColumns: repeat(
            col || AUTO_FIT,
            colWidth,
            colMinWidth,
            colMaxWidth
          ),
        }),

    alignItems,
    justifyContent,
    gap,
    ...otherStyles,
  };

  return (
    <CellComponent style={style} {...rest}>
      {children}
    </CellComponent>
  );
};

export default Grid;
