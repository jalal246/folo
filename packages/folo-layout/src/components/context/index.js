export { GridConsumer, GridProvider } from "./GridContext";
