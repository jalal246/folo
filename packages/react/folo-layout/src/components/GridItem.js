import React from "react";
import positionStore from "../positionStore";

import {
  CENTER,
  SPACE_BETWEEN,
  ROW,
  COLUMN,
  DISPLAY_FLEX,
  STRETCH,
} from "../constants";

function location(from, to) {
  if (from !== null && to !== null) {
    return `${from} / ${to}`;
  }
  return `${from}`;
}

/**
 * For implicit grid
 * Takes column and row number
 * to inform the grid parent of total positions
 *
 * It collects numbers report it to Grid
 */
const GridItem = (props) => {
  const {
    component: CellComponent = "div",

    row = null,
    toRow = null,

    col = 0,
    toCol = null,

    isCenter = false,
    isHorizontal = true,

    style: {
      display = DISPLAY_FLEX,

      flexDirection: fDirection,
      alignItems: aItems,
      ...otherStyle
    } = {},

    id,
    children,

    ...rest
  } = props;

  const calculatedPosition = positionStore.autoPosition({
    key: id || `${new Date().getTime()}`,
    row,
    toRow,
  });

  const container = {
    display,

    ...(isHorizontal
      ? {
          flexDirection: fDirection || ROW,
          alignItems: aItems || CENTER,
        }
      : {
          flexDirection: fDirection || COLUMN,
          alignItems: aItems || STRETCH,
        }),

    justifyContent: isCenter ? CENTER : SPACE_BETWEEN,

    gridRow: location(calculatedPosition, toRow),
    gridColumn: location(col, toCol),

    ...otherStyle,
  };

  return (
    <CellComponent style={container} {...rest}>
      {children}
    </CellComponent>
  );
};

export default GridItem;
