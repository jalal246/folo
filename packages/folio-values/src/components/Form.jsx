import React from "react";
import PropTypes from "prop-types";

import withContext from "@folio/withContext";

import { ValuesConsumer, ValuesProvider } from "./context";

const propTypes = {
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  onSubmit: PropTypes.func,
  getContextValues: PropTypes.func
};

const defaultProps = {
  component: "form",
  onSubmit() {},
  getContextValues() {}
};

class Form extends React.PureComponent {
  onSubmit = e => {
    e.preventDefault();
    const { getContextValues, onSubmit } = this.props;
    onSubmit(e, getContextValues());
  };

  render() {
    const {
      component: FormComponent,
      onSubmit,
      getContextValues,
      children,
      ...other
    } = this.props;
    return (
      <ValuesProvider>
        <FormComponent onSubmit={this.onSubmit} {...other}>
          {children}
        </FormComponent>
      </ValuesProvider>
    );
  }
}

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;

export { Form as PureForm };

export default withContext({
  Component: Form,
  Consumer: ValuesConsumer,
  contextProps: ["getContextValues"]
});
