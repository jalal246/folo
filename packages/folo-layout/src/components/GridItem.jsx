import React, { PureComponent } from "react";
import PropTypes from "prop-types";

import withcontext from "@folo/withcontext";

import { GridConsumer } from "./context";

import {
  CENTER,
  SPACE_BETWEEN,
  ROW,
  COLUMN,
  DISPLAY_FLEX,
  STRETCH
} from "../constants";

function location(from, to) {
  if (from !== null && to !== null) {
    return `${from} / ${to}`;
  }
  return `${from}`;
}

const propTypes = {
  component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),

  row: PropTypes.number,
  toRow: PropTypes.number,

  col: PropTypes.number,
  toCol: PropTypes.number,
  isCenter: PropTypes.bool,

  style: PropTypes.objectOf(PropTypes.string),

  autoPosition: PropTypes.func.isRequired,

  isHorizontal: PropTypes.bool,
  children: PropTypes.node
};

const defaultProps = {
  component: "div",

  row: null,
  toRow: null,

  col: 0,
  toCol: null,
  isCenter: false,

  style: {},

  isHorizontal: true,
  children: null
};

/**
 * For implicit grid
 * Takes column and row number
 * to inform the grid parent of total positions
 *
 * It collects numbers report it to Grid
 */
class GridItem extends PureComponent {
  state = {
    key: new Date().getTime()
  };

  render() {
    const {
      component: CellComponent,

      row,
      toRow,

      col,
      toCol,

      isCenter,
      isHorizontal,

      style: {
        display = DISPLAY_FLEX,

        flexDirection: fDirection,
        alignItems: aItems,
        ...otherStyle
      },

      autoPosition,

      children,

      ...rest
    } = this.props;

    const { key } = this.state;

    // console.log('GridItem updated');
    const calculatedPosition = autoPosition({ key, row, toRow });

    const container = {
      display,

      ...(isHorizontal
        ? {
            flexDirection: fDirection || ROW,
            alignItems: aItems || CENTER
          }
        : {
            flexDirection: fDirection || COLUMN,
            alignItems: aItems || STRETCH
          }),

      justifyContent: isCenter ? CENTER : SPACE_BETWEEN,

      gridRow: location(calculatedPosition, toRow),
      gridColumn: location(col, toCol),

      ...otherStyle
    };

    return (
      <CellComponent style={container} {...rest}>
        {children}
      </CellComponent>
    );
  }
}

GridItem.propTypes = propTypes;
GridItem.defaultProps = defaultProps;

export { GridItem as PureGridItem };

export default withcontext({
  Component: GridItem,
  Consumer: GridConsumer,
  contextProps: ["autoPosition"]
});
