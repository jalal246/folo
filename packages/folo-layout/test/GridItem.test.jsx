import React from "react";
import { render, cleanup } from "react-testing-library";

import { PureGridItem } from "../src/components/GridItem";

const autoPosition = jest.fn(() => 0);

const TESTID = "testidforGridItem";

const CENTER = "center";
const ROW = "row";
const COLUMN = "column";
const DISPLAY_FLEX = "flex";
const SPACE_BETWEEN = "space-between";
const STRETCH = "stretch";

const MyGridItem = (props = {}) => (
  <PureGridItem data-testid={TESTID} autoPosition={autoPosition} {...props} />
);

afterEach(cleanup);

describe("GridItem", () => {
  describe("style", () => {
    it("returns default style", () => {
      const { baseElement, getByTestId } = render(<MyGridItem />);
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        gridRow: "0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("returns flexDirection column when it is isHorizontal = false", () => {
      const { baseElement, getByTestId } = render(
        <MyGridItem isHorizontal={false} />
      );
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: STRETCH,
        flexDirection: COLUMN,
        gridRow: "0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("overrides styles", () => {
      const { baseElement, getByTestId } = render(
        <MyGridItem
          style={{
            flexDirection: COLUMN,
            alignItems: STRETCH
          }}
        />
      );
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: STRETCH,
        flexDirection: COLUMN,
        gridRow: "0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("returns center", () => {
      const { baseElement, getByTestId } = render(<MyGridItem isCenter />);
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        gridRow: "0",
        gridColumn: "0",
        justifyContent: CENTER
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });

  describe("gridRow", () => {
    it("returns gridRow with toRow and alignItems center", () => {
      const { baseElement, getByTestId } = render(<MyGridItem toRow={6} />);
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        gridRow: "0 / 6",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("returns gridRow with toRow =0", () => {
      const { baseElement, getByTestId } = render(<MyGridItem toRow={0} />);
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        gridRow: "0 / 0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("returns gridRow with toRow =0", () => {
      const { baseElement, getByTestId } = render(<MyGridItem toRow={0} />);
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        gridRow: "0 / 0",
        gridColumn: "0",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });

  describe("gridColumn", () => {
    it("returns column position with when col is provided", () => {
      const { baseElement, getByTestId } = render(<MyGridItem col={10} />);
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        gridRow: "0",
        gridColumn: "10",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });

    it("returns column position with justifyContent, when toCol is provided", () => {
      const { baseElement, getByTestId } = render(<MyGridItem toCol={10} />);
      const { style } = getByTestId(TESTID);

      const expected = {
        display: DISPLAY_FLEX,
        alignItems: CENTER,
        flexDirection: ROW,
        gridRow: "0",
        gridColumn: "0 / 10",
        justifyContent: SPACE_BETWEEN
      };

      expect(baseElement).toMatchSnapshot();
      expect(style).toMatchObject(expected);
    });
  });
});
