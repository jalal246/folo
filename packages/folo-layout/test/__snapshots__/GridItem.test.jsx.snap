// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridItem gridColumn returns column position with justifyContent, when toCol is provided 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: row; align-items: center; justify-content: space-between;"
    />
  </div>
</body>
`;

exports[`GridItem gridColumn returns column position with when col is provided 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: row; align-items: center; justify-content: space-between;"
    />
  </div>
</body>
`;

exports[`GridItem gridRow returns gridRow with toRow =0 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: row; align-items: center; justify-content: space-between;"
    />
  </div>
</body>
`;

exports[`GridItem gridRow returns gridRow with toRow =0 2`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: row; align-items: center; justify-content: space-between;"
    />
  </div>
</body>
`;

exports[`GridItem gridRow returns gridRow with toRow and alignItems center 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: row; align-items: center; justify-content: space-between;"
    />
  </div>
</body>
`;

exports[`GridItem style overrides styles 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: column; align-items: stretch; justify-content: space-between;"
    />
  </div>
</body>
`;

exports[`GridItem style returns center 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: row; align-items: center; justify-content: center;"
    />
  </div>
</body>
`;

exports[`GridItem style returns default style 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: row; align-items: center; justify-content: space-between;"
    />
  </div>
</body>
`;

exports[`GridItem style returns flexDirection column when it is isHorizontal = false 1`] = `
<body>
  <div>
    <div
      data-testid="testidforGridItem"
      style="display: flex; flex-direction: column; align-items: stretch; justify-content: space-between;"
    />
  </div>
</body>
`;
